From 88d9ce22058adcd18fa0ea39c893b7dd6861cec1 Mon Sep 17 00:00:00 2001
From: Christian Brabandt <cb@256bit.org>
Date: Sat, 12 Sep 2026 17:32:08 +0000
Subject: [PATCH] patch 9.2.1090: [security]: ex command injection in sign jump

Problem:  [security]: ex command injection in sign jump (skeletonsec)
Solution: Use do_ecmd() directly instead of constructing a command line

Github Security Advisory:
https://github.com/vim/vim/security/advisories/GHSA-x9jf-rjm6-vxqh

Supported by AI.

Signed-off-by: Christian Brabandt <cb@256bit.org>
---
 src/sign.c                 | 10 +++-------
 src/testdir/test_signs.vim | 22 ++++++++++++++++++++++
 2 files changed, 25 insertions(+), 7 deletions(-)

diff --git a/src/sign.c b/src/sign.c
index fd373b5c3..562529eba 100644
--- a/src/sign.c
+++ b/src/sign.c
@@ -1265,19 +1265,15 @@ sign_jump(int sign_id, char_u *sign_group, buf_T *buf)
     }
     else
     {			// ... not currently in a window
-	char_u	*cmd;
-
 	if (buf->b_fname == NULL)
 	{
 	    emsg(_("E934: Cannot jump to a buffer that does not have a name"));
 	    return -1;
 	}
-	cmd = alloc(STRLEN(buf->b_fname) + 25);
-	if (cmd == NULL)
+	setpcmark();
+	if (do_ecmd(buf->b_fnum, NULL, NULL, NULL, lnum,
+		    buf_hide(curbuf) ? ECMD_HIDE : 0, curwin) == FAIL)
 	    return -1;
-	sprintf((char *)cmd, "e +%ld %s", (long)lnum, buf->b_fname);
-	do_cmdline_cmd(cmd);
-	vim_free(cmd);
     }
 # ifdef FEAT_FOLDING
     foldOpenCursor();
diff --git a/src/testdir/test_signs.vim b/src/testdir/test_signs.vim
index 738eebf4d..ac8734654 100644
--- a/src/testdir/test_signs.vim
+++ b/src/testdir/test_signs.vim
@@ -2012,4 +2012,26 @@ func Test_sign_funcs_multi()
   call delete("Xsign")
 endfunc
 
+" A buffer name containing a bar must not be interpreted as an Ex command
+" separator when jumping to a sign in a buffer that is not displayed.
+func Test_sign_jump_name_with_bar()
+  CheckFeature signs
+
+  let bufnr = bufadd('Xsign|call setline(1, "PWNED")')
+  call bufload(bufnr)
+  call setbufline(bufnr, 1, ['one', 'two', 'three'])
+
+  sign define sjTest text=x
+  call sign_place(1, '', 'sjTest', bufnr, #{lnum: 2})
+
+  call sign_jump(1, '', bufnr)
+
+  call assert_equal(bufnr, bufnr('%'))
+  call assert_equal(2, line('.'))
+  call assert_equal('two', getline(2))
+
+  call sign_undefine('sjTest')
+  exe 'bwipe! ' .. bufnr
+endfunc
+
 " vim: shiftwidth=2 sts=2 expandtab
